File: 129.sum-root-to-leaf-numbers.py
#
# @lc app=leetcode id=129 lang=python3
#
# [129] Sum Root to Leaf Numbers
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

#  太丑，想想怎么写好看
class Solution:
    def __init__(self):
        self.total = 0
        self.current = 0
    def sumNumbers(self, root: TreeNode) -> int:
        if not root:
            return 0
        self.current = root.val
        self.helper(root)
        return self.total
        
    def helper(self,root):
        if not root.left and not root.right:
            self.total += self.current
            return 

        self.current *= 10
        if root.left:
            self.current += root.left.val
            self.helper(root.left)
            self.current -= root.left.val

        if root.right:
            self.current += root.right.val
            self.helper(root.right)
            self.current -= root.right.val
        self.current = self.current // 10
        
        
# @lc code=end

File: golang/129.sum-root-to-leaf-numbers.go
/*
 * @lc app=leetcode.cn id=129 lang=golang
 *
 * [129] Sum Root to Leaf Numbers
 */
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
	if root == nil {
		return 0
	}
	ret := 0
	collect := func(delta int) {
		ret += delta
	}
	cur := 0
	sumNumberHelper(root, cur, collect)
	return ret

}

func sumNumberHelper(root *TreeNode, cur int, collect func(int)) {
	if root.Left == nil && root.Right == nil {
		collect(cur*10 + root.Val)
		return
	}
	cur = cur*10 + root.Val
	if root.Left != nil {
		sumNumberHelper(root.Left, cur, collect)
	}
	if root.Right != nil {
		sumNumberHelper(root.Right, cur, collect)
	}

}

// @lc code=end
